File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.py

class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        s = sum(nums)
        
        if s & 1 == 1:
            return False

        target = s // 2

        dp = [False] * (target+1)
        dp[0] = True
        
        for num in nums:
            for i in range(target, 0, -1):
                if i >= num:
                    dp[i] = dp[i] or dp[i-num]

        return dp[target]

File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.cpp
class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int total = 0;
        for (int n : nums) {
            total += n;
        }
        
        if (total % 2 == 1) return false;
        total /= 2;
        
        vector<bool> dp(total+1, false);
        dp[0] = true;
        
        for (int n : nums) {
            for (int i = total; i >= n; i--) {
                dp[i] = dp[i] || dp[i-n];
            }
            if (dp[total]) return true;
        }
        
        return dp[total];
    }
};